File: FledgeJava/app/src/main/java/com/example/adservices/samples/fledge/sampleapp/AdSelectionClient.java
/*
 * Copyright (C) 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.example.adservices.samples.fledge.sampleapp;

import android.adservices.adselection.AdSelectionConfig;
import android.adservices.adselection.AdSelectionManager;
import android.adservices.adselection.AdSelectionOutcome;
import android.adservices.adselection.ReportImpressionRequest;
import android.adservices.exceptions.AdServicesException;
import android.content.Context;
import android.net.Uri;
import android.os.OutcomeReceiver;
import android.util.Log;
import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executor;
import java.util.function.Consumer;
import java.util.function.Supplier;
import java.util.stream.Collectors;
import org.json.JSONObject;

/**
 * Wrapper for the FLEDGE Ad Selection API. This wrapper is opinionated and makes several
 * choices such as running impression report immediately after every successful ad auction or leaving
 * the ad signals empty to limit the complexity that is exposed the user.
 */
@RequiresApi(api = 33)
public class AdSelectionClient {

  private final List<String> mBuyers;
  private final String mSeller;
  private final Supplier<Uri> mDecisionUrlSupplier;
  private final AdSelectionManager mAdManager;
  private final Executor mExecutor;

  /**
   * Initializes the ad selection client with a specific seller, list of buyers, and decision
   * endpoint.
   * @param buyers A list of buyers for the auction.
   * @param seller The name of the seller for the auction
   * @param decisionUrlSupplier A supplier that gives theURL to retrieve the seller scoring and
   * reporting logic from
   * @param context The application context.
   * @param executor An executor to use with the FLEDGE API calls.
   */
  public AdSelectionClient(List<String> buyers, String seller, Supplier<Uri> decisionUrlSupplier,
      Context context, Executor executor) {
    mBuyers = buyers;
    mSeller = seller;
    mDecisionUrlSupplier = decisionUrlSupplier;
    mAdManager = context.getSystemService(AdSelectionManager.class);
    mExecutor = executor;
  }

  /**
   * Runs add selection and passes a string describing it's status to the input receivers. If ad
   * selection succeeds, also report impressions.
   * @param statusReceiver A consumer function that is run after ad selection and impression reporting
   * with a string describing how the auction and reporting went.
   * @param renderUrlReceiver A consumer function that is run after ad selection with a message describing the render URL
   * or lack thereof.
   */
  public void runAdSelection(Consumer<String> statusReceiver, Consumer<String> renderUrlReceiver) {
    try {
      AdSelectionConfig config = generateAdSelectionConfig();
      OutcomeReceiver<AdSelectionOutcome, AdServicesException> adSelectionReceiver =
          new OutcomeReceiver<AdSelectionOutcome, AdServicesException>() {
            @Override
            public void onResult(@NonNull AdSelectionOutcome adSelectionOutcome) {
              statusReceiver.accept("Ran ad selection");
              renderUrlReceiver.accept("Would display ad from " + adSelectionOutcome.getRenderUrl());


              reportImpression(adSelectionOutcome.getAdSelectionId(), config, statusReceiver);
            }

            @Override
            public void onError(@NonNull AdServicesException e) {
              statusReceiver.accept("Error when running ad selection: " + e.getMessage());
              renderUrlReceiver.accept("Ad selection failed -- no ad to display");
              Log.e(MainActivity.TAG, "Exception during ad selection", e);
            }
          };
      mAdManager.runAdSelection(config, mExecutor, adSelectionReceiver);
    } catch (Exception e) {
      statusReceiver.accept("Got the following exception when trying to run ad selection: " + e);
      renderUrlReceiver.accept("Ad selection failed -- no ad to display");
      Log.e(MainActivity.TAG, "Exception calling runAdSelection", e);
    }

  }

  /**
   * Generates an ad selection config based on the fields in this class.
   * @return The generated ad selection config
   */
  private AdSelectionConfig generateAdSelectionConfig() {
    return new AdSelectionConfig.Builder()
        .setSeller(mSeller)
        .setDecisionLogicUrl(mDecisionUrlSupplier.get())
        .setCustomAudienceBuyers(mBuyers)
        .setAdSelectionSignals(new JSONObject().toString())
        .setSellerSignals(new JSONObject().toString())
        .setPerBuyerSignals(mBuyers.stream()
            .collect(Collectors.toMap(buyer -> buyer, buyer -> new JSONObject().toString())))
        .setContextualAds(new ArrayList<>())
        .build();
  }

  /**
   * Helper function of {@link #runAdSelection}. Runs impression reporting.
   *
   * @param adSelectionId The auction to report impression on.
   * @param statusReceiver A consumer function that is run after impression reporting
   * with a string describing how the auction and reporting went.
   */
  private void reportImpression(long adSelectionId, AdSelectionConfig config, Consumer<String> statusReceiver) {
    ReportImpressionRequest request = new ReportImpressionRequest.Builder()
        .setAdSelectionConfig(config)
        .setAdSelectionId(adSelectionId)
        .build();
    OutcomeReceiver<Void, AdServicesException> impressionReceiver = new OutcomeReceiver<Void, AdServicesException>() {
      @Override
      public void onResult(@NonNull Void unused) {
        statusReceiver.accept("Reported impressions from ad selection");
      }

      @Override
      public void onError(@NonNull AdServicesException error) {
        statusReceiver.accept("Error when reporting impressions: " + error.getMessage());
        Log.e(MainActivity.TAG, error.toString(), error);
      }
    };
    mAdManager.reportImpression(request, mExecutor, impressionReceiver);
  }

}

File: Fledge/FledgeKotlin/app/src/main/java/com/example/adservices/samples/fledge/adselection/AdSelectionClient.kt
/*
 * Copyright (C) 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.example.adservices.samples.fledge.adselection

import android.util.Log
import androidx.privacysandbox.ads.adservices.adselection.AdSelectionConfig
import androidx.privacysandbox.ads.adservices.adselection.AdSelectionManager
import androidx.privacysandbox.ads.adservices.adselection.AdSelectionOutcome
import androidx.privacysandbox.ads.adservices.adselection.GetAdSelectionDataOutcome
import androidx.privacysandbox.ads.adservices.adselection.GetAdSelectionDataRequest
import androidx.privacysandbox.ads.adservices.adselection.PersistAdSelectionResultRequest
import androidx.privacysandbox.ads.adservices.adselection.ReportEventRequest
import androidx.privacysandbox.ads.adservices.adselection.ReportImpressionRequest
import androidx.privacysandbox.ads.adservices.adselection.UpdateAdCounterHistogramRequest
import androidx.privacysandbox.ads.adservices.common.ExperimentalFeatures
import androidx.privacysandbox.ads.adservices.common.FrequencyCapFilters
import com.example.adservices.samples.fledge.app.FledgeApplication.Companion.TAG
import com.example.adservices.samples.fledge.app.FledgeApplication.Companion.appContext
import java.util.function.Consumer
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class AdSelectionClient {
  private val adSelectionManager: AdSelectionManager = AdSelectionManager.obtain(appContext)!!

  /**
   * Runs ad selection.
   */
  suspend fun selectAds(
    adSelectionConfig: AdSelectionConfig,
    statusReceiver: Consumer<String>
  ): AdSelectionOutcome {
    return try {
      withContext(Dispatchers.IO) {
        val adSelectionOutcome = adSelectionManager.selectAds(adSelectionConfig)
        statusReceiver.accept("Ad selection successful for " + adSelectionOutcome.adSelectionId)
        adSelectionOutcome
      }
    } catch (e: Exception) {
      statusReceiver.accept("Exception calling runAdSelection: $e")
      Log.e(TAG, "Exception calling runAdSelection", e)
      throw e
    }
  }

  /**
   * Runs impression reporting.
   */
  @OptIn(ExperimentalFeatures.Ext8OptIn::class)
  suspend fun reportImpression(
    request: ReportImpressionRequest,
    statusReceiver: Consumer<String>,
    onSuccess: (suspend () -> Unit)? = null
  ) {
    try {
      withContext(Dispatchers.IO) {
        adSelectionManager.reportImpression(request)
      }
      statusReceiver.accept("Reported impression from ad selection")
      Log.v(TAG, "Reported impression from ad selection")
      onSuccess?.invoke()
    } catch (e: Exception) {
      statusReceiver.accept("Exception calling reportImpression: ${e.message}")
      Log.e(TAG, "Exception calling reportImpression: ${e.message}")
    }
  }

  /**
   * Runs event reporting.
   */
  @OptIn(ExperimentalFeatures.Ext8OptIn::class)
  suspend fun reportEvent(
    request: ReportEventRequest,
    statusReceiver: Consumer<String>,
  ) {
    try {
      withContext(Dispatchers.IO) {
        adSelectionManager.reportEvent(request)
      }
      statusReceiver.accept("Reported event with ${request.eventKey} key from ad selection")
    } catch (e: Exception) {
      statusReceiver.accept("Exception calling reportEvent with ${request.eventKey} key and ${request.eventData} data: ${e.message}")
      Log.e(
        TAG,
        "Exception calling reportEvent with ${request.eventKey} key and ${request.eventData} data: ${e.message}"
      )
    }
  }

  /**
   * Updates the counter histogram for an ad.
   */
  @OptIn(ExperimentalFeatures.Ext8OptIn::class)
  suspend fun updateAdCounterHistogram(
    request: UpdateAdCounterHistogramRequest,
    statusReceiver: Consumer<String>
  ) {

    try {
      withContext(Dispatchers.IO) {
        adSelectionManager.updateAdCounterHistogram(request)
      }
      statusReceiver.accept(
        "Updated ad counter histogram with ${fCapEventToString(request.adEventType)} event for adtech ${request.callerAdTech}"
      )
    } catch (e: Exception) {
      statusReceiver.accept(
        "Exception calling updateAdCounterHistogram with ${fCapEventToString(request.adEventType)} event for adtech ${request.callerAdTech}: ${e.message}"
      )
      Log.e(
        TAG,
        "Exception calling updateAdCounterHistogram with ${fCapEventToString(request.adEventType)} event for adtech ${request.callerAdTech}: ${e.message}"
      )
    }
  }

  /**
   * Invokes the `getAdSelectionData` method of [AdSelectionManager], and returns an GetAdSelectionDataOutcome.
   */
  @ExperimentalFeatures.Ext10OptIn
  suspend fun getAdSelectionData(
    request: GetAdSelectionDataRequest,
    statusReceiver: Consumer<String>
  ): GetAdSelectionDataOutcome {
    return try {
      withContext(Dispatchers.IO) {
        adSelectionManager.getAdSelectionData(request)
      }
    } catch (e: Exception) {
      statusReceiver.accept("Got the following exception when trying to run getAdSelectionData: $e")
      Log.e(TAG, "Exception calling getAdSelectionData: ${e.message}", e)
      throw e
    }
  }

  /**
   * Invokes the `persistAdSelectionResult` method of [AdSelectionManager], and returns an AdSelectionOutcome.
   */
  @OptIn(ExperimentalFeatures.Ext10OptIn::class)
  suspend fun persistAdSelectionResult(
    request: PersistAdSelectionResultRequest,
    statusReceiver: Consumer<String>
  ): AdSelectionOutcome {
    return try {

      withContext(Dispatchers.IO) {
        val adSelectionOutcome = adSelectionManager.persistAdSelectionResult(request)
        statusReceiver.accept("Auction Result is persisted for ${adSelectionOutcome.adSelectionId}")
        Log.v(TAG, "Auction Result is persisted for ${adSelectionOutcome.adSelectionId}")
        adSelectionOutcome
      }
    } catch (e: Exception) {
      statusReceiver.accept("Exception calling persistAdSelectionResult: $e")
      Log.e(TAG, "Exception calling persistAdSelectionResult: ${e.message}", e)
      throw e
    }
  }

  @OptIn(ExperimentalFeatures.Ext8OptIn::class)
  private fun fCapEventToString(eventType: Int): String {
    val result: String =
      when (eventType) {
        FrequencyCapFilters.AD_EVENT_TYPE_WIN -> "win"
        FrequencyCapFilters.AD_EVENT_TYPE_CLICK -> "click"
        FrequencyCapFilters.AD_EVENT_TYPE_IMPRESSION -> "impression"
        FrequencyCapFilters.AD_EVENT_TYPE_VIEW -> "view"
        else -> "unknown"
      }
    return result
  }
}
